realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/AuthTests.java

```diff
@@ -266,4 +270,130 @@ public void changePassword_throwWhenUserIsLoggedOut() {
         user.changePassword("new-password");
     }
 
+    // Cached instances of RealmConfiguration should not be allowed to be used if the user is no longer valid
+    @Test
+    public void cachedInstanceShouldThrowIfUserBecomeInvalid() throws InterruptedException {
+        String username = UUID.randomUUID().toString();
+        String password = "password";
+
+        SyncCredentials credentials = SyncCredentials.usernamePassword(username, password, true);
+        SyncUser user = SyncUser.login(credentials, Constants.AUTH_URL);
+        final RealmConfiguration configuration = new SyncConfiguration.Builder(user, Constants.USER_REALM).build();
+        Realm.getInstance(configuration);
+
+        user.logout();
+        assertFalse(user.isValid());
+
+        final CountDownLatch backgroundThread = new CountDownLatch(1);
+        final IllegalStateException[] backgroundException = new IllegalStateException[1];
+        // Should throw when using the invalid configuration form a different thread
+        new Thread() {
+            @Override
+            public void run() {
+                try {
+                    Realm.getInstance(configuration);
+                } catch (IllegalStateException expected) {
+                    backgroundException[0] = expected;
```

Comment:
you can assert here, `assertNotNull(backgroundException[0]);`. JUnit4 will fail the test when assertion fails in the bg thread.